staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/marshal_test.go

```diff
@@ -148,3 +522,558 @@ func TestJSONSchemaPropsOrArrayMarshalJSON(t *testing.T) {
+				Schema: &JSONSchemaProps{},
+			},
+		},
+		{
+			name:  "empty array value",
+			input: "80", // []
+			expected: &JSONSchemaPropsOrStringArray{
+				Property: []string{},
+			},
+		},
+		{
+			name:  "props value",
+			input: "A1447479706546737472696E67", // {"type":"string"}
+			expected: &JSONSchemaPropsOrStringArray{
+				Schema: &JSONSchemaProps{Type: "string"},
+			},
+		},
+		{
+			name:  "array value",
+			input: "8146737472696E67", // ["string"]
+			expected: &JSONSchemaPropsOrStringArray{
+				Property: []string{"string"},
+			},
+		},
+		{
+			name:     "wrong type",
+			input:    "182A", // 42
+			expected: &JSONSchemaPropsOrStringArray{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			input, err := hex.DecodeString(tc.input)
+			if err != nil {
+				t.Fatal(err)
+			}
+			var actual *JSONSchemaPropsOrStringArray
+			if err := cbor.Unmarshal(input, &actual); err != nil {
+				t.Fatal("unexpected error:", err)
+			}
+			if !reflect.DeepEqual(tc.expected, actual) {
+				t.Fatal(cmp.Diff(tc.expected, actual))
+			}
+		})
+	}
+
+	t.Run("direct invocation", func(t *testing.T) {
+		testCases := []struct {
+			name  string
+			input []byte
+		}{
+			{
+				name:  "null",
+				input: []byte{cborNullValue},
+			},
+			{
+				name:  "zero len",
+				input: []byte{},
+			},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				expected := JSONSchemaPropsOrStringArray{}
+				actual := JSONSchemaPropsOrStringArray{
+					Property: []string{"override me"},
+					Schema:   &JSONSchemaProps{Type: "override me"},
+				}
+				err := actual.UnmarshalCBOR(tc.input)
+				if err != nil {
+					t.Fatal("unexpected error:", err)
+				}
+				if !reflect.DeepEqual(expected, actual) {
+					t.Fatal(cmp.Diff(expected, actual))
+				}
+			})
+		}
+	})
+
+}
+
+func TestJSONMarshalJSON(t *testing.T) {
+	tastCases := []struct {
+		name     string
+		input    *JSON
+		expected string
+	}{
+		{
+			name:     "nil",
+			input:    nil,
+			expected: `null`,
+		},
+		{
+			name:     "empty",
+			input:    &JSON{},
+			expected: `null`,
+		},
+		{
+			name:     "string",
+			input:    &JSON{Raw: []byte(`"string"`)},
+			expected: `"string"`,
+		},
+		{
+			name:     "number",
+			input:    &JSON{Raw: []byte(`42.0`)},
+			expected: `42.0`,
+		},
+		{
+			name:     "bool",
+			input:    &JSON{Raw: []byte(`true`)},
+			expected: `true`,
+		},
+		{
+			name:     "array",
+			input:    &JSON{Raw: []byte(`[1,2,3]`)},
+			expected: `[1,2,3]`,
+		},
+		{
+			name:     "map",
+			input:    &JSON{Raw: []byte(`{"foo":"bar"}`)},
+			expected: `{"foo":"bar"}`,
+		},
+	}
+	for _, tc := range tastCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual, err := json.Marshal(tc.input)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if tc.expected != string(actual) {
+				t.Fatal(cmp.Diff(tc.expected, actual))
+			}
+		})
+	}
+}
+
+func TestJSONUnmarshalJSON(t *testing.T) {
```

Comment:
A test case for `[]byte("null")` would be interesting given the inconsistent handling in this area.